src/main/java/org/opensearch/neuralsearch/util/prune/PruneUtils.java

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.util.prune;
+
+import org.opensearch.common.collect.Tuple;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.PriorityQueue;
+
+/**
+ * Utility class providing methods for prune sparse vectors using different strategies.
+ * Prune helps reduce the dimensionality of sparse vectors by removing less significant elements
+ * based on various criteria.
+ */
+public class PruneUtils {
+    public static final String PRUNE_TYPE_FIELD = "prune_type";
+    public static final String PRUNE_RATIO_FIELD = "prune_ratio";
+
+    /**
+     * Prunes a sparse vector by keeping only the top K elements with the highest values.
+     *
+     * @param sparseVector The input sparse vector as a map of string keys to float values
+     * @param k The number of top elements to keep
+     * @param requiresPrunedEntries Whether to return pruned entries
+     * @return A tuple containing two maps: the first with top K elements, the second with remaining elements (or null)
+     */
+    private static Tuple<Map<String, Float>, Map<String, Float>> pruneByTopK(
+        Map<String, Float> sparseVector,
+        float k,
+        boolean requiresPrunedEntries
+    ) {
+        PriorityQueue<Map.Entry<String, Float>> pq = new PriorityQueue<>((a, b) -> Float.compare(a.getValue(), b.getValue()));
+
+        for (Map.Entry<String, Float> entry : sparseVector.entrySet()) {
+            if (pq.size() < (int) k) {
+                pq.offer(entry);
+            } else if (entry.getValue() > pq.peek().getValue()) {
+                pq.poll();
+                pq.offer(entry);
+            }
+        }
+
+        Map<String, Float> highScores = new HashMap<>();
+        Map<String, Float> lowScores = requiresPrunedEntries ? new HashMap<>(sparseVector) : null;
+
+        while (!pq.isEmpty()) {
+            Map.Entry<String, Float> entry = pq.poll();
+            highScores.put(entry.getKey(), entry.getValue());
+            if (requiresPrunedEntries) {
+                lowScores.remove(entry.getKey());
```

Comment:
can we check if `lowScores` is not null instead of checking the `requiresPrunedEntries` flag? with current code initialization and check code are disconnected, later this code can evolve and meaning of `requiresPrunedEntries` can change.

Reply:
ack